server/authentication.js

```diff
@@ -8,42 +13,220 @@ function enforceHTTPS(request, response, next) {
+    });
+}
+
+function isOnWhitelist(pool, email){
+  const whitelistSQL = 'SELECT * FROM whitelist WHERE email=$1 ORDER BY id ASC LIMIT 1';
+  const whitelistValues = [email];
+  
+  return pool.query(whitelistSQL, whitelistValues)
+    .then(results => Promise.resolve(results.rowCount === 0));
+}
+
+function createLinkAndEmail(pool, mailgunEnv, email, domain) {
+  return insertLink(pool,email, domain)
+    .then(link => emailLink(mailgunEnv, email, link));
+}
+
+function insertLink(pool, email, domain) {
+  const linkToken = uuid.v4();
+  //TODO: what should link look like???
+  const link = `${domain}/login_from_email?${qs.stringify({linkToken})}`;
+
+  // Insert link into database
+  const linkSQL = `INSERT INTO links(email, link, timestampz) VALUES ($1, $2, $3)`;
+  const now = new Date();
+  const linkValues = [email, linkToken, now];
+  return pool.query(linkSQL, linkValues)
+    .then(results => link)
+    .catch(err => {
+      console.log('query returned err: ', err);
+      console.log({ error:err });
+    });
+}
+
+function emailLink(mailgunEnv, email, link) {
+  const linkText = link;
+  const loginlinkFilename = path.join(__dirname,'game/emails/loginlink.html.mustache');
+  const html = renderEmail(loginlinkFilename,{linkText});
+
+  const info = {
+    toEmail: email,
+    fromEmail: 'swipe-right-bot@tsl.mit.edu',
+    subject: 'Swipe Right for CS: Login Link'
+  };
+
+  if (process.env.NODE_ENV !== 'production') {
+    if (process.env.NODE_ENV === 'development') {
+      console.log('No emailing except for in production mode. Go to the following link to move forward.');
+      console.log(link);
+    }
+    return Promise.resolve();
+  }
+
+  return new Promise((resolve, reject) => {
+    sendEmail(mailgunEnv, info, html, (err, mailgunResponse) => {
+      if (err) {
+        console.log("Mailgun request (err):\n  " + JSON.stringify(err, null, 2));
+        return reject(err);
+      }
+      console.log("Mailgun returned:\n  " + JSON.stringify(mailgunResponse, null, 2));
+      return resolve();   
+    });
+  });
+}
+
+// Endpoint to check link from researchers' email
+// Confirm email link is valid and generates token 
+// for user to access data. Adds token to 'tokens' database
+// 
+function emailLinkEndpoint(pool, request, response){
+  const link = request.body['link'];
+  const email = request.body['email'];
+  checkLink(pool, email, link)
+    .then(linkAuthorized => {
+      if (linkAuthorized) {
+        console.log('Link Authorized');
+
+        //TODO: feel like these promises are not done right...shouldn't have to nest things like this
+        generateToken(pool, email)
+          .then(results => {
+            const token = results;
+            response.set('Content-Type', 'application/json');
+            response.json({ 
+              token: token
+            });
+            return response.status(200).end();
+          });
+      }
+      else {
+        console.log('bad link');
```

Comment:
done! logging emails too since both contribute

Reply:
@keving17 That's a good thought!  Since emails are more sensitive and identify individual users, it's best to avoid logging them when you can.  Here, since we can use the token to find what email it's tied to, could we remove the email and still be able to debug later?

Reply:
I think you make a good point. We should be able to debug with just the link so I'll remove the email.